src/session/session_api.c

```diff
@@ -1583,55 +1584,103 @@ __wt_session_range_truncate(
     trunc_info->stop = stop;
     trunc_info->orig_start_key = orig_start_key;
     trunc_info->orig_stop_key = orig_stop_key;
-    if (uri != NULL)
-        trunc_info->uri = uri;
-    else if (start != NULL)
-        trunc_info->uri = start->internal_uri;
-    else {
-        WT_ASSERT(session, stop != NULL);
-        trunc_info->uri = stop->internal_uri;
-    }
+    trunc_info->uri = actual_uri;
+
+    /*
+     * We would like to use bounded cursors if possible, so check whether they are supported. We
+     * don't currently support bounded cursors for fixed column stores. For now, also fall back to
+     * not using bounded cursors for complex types, such as tables with column groups and indexes.
+     * Those tables may or may not support bounded cursors and it is not easy to check at this
+     * abstraction level.
+     */
+    if (CUR2BT(start) == NULL || CUR2BT(start)->type == BTREE_COL_FIX)
+        supports_bounds = false;
+    if (stop != NULL && (CUR2BT(stop) == NULL || CUR2BT(stop)->type == BTREE_COL_FIX))
+        supports_bounds = false;
 
     /*
      * Truncate does not require keys actually exist so that applications can discard parts of the
      * object's name space without knowing exactly what records currently appear in the object. For
-     * this reason, do a search-near, rather than a search. Additionally, we have to correct after
-     * calling search-near, to position the start/stop cursors on the next record greater than/less
-     * than the original key. If we fail to find a key in a search-near, there are no keys in the
-     * table. If we fail to move forward or backward in a range, there are no keys in the range. In
-     * either of those cases, we're done.
+     * this reason, position the start/stop cursors to the actual first/last keys in the truncation
+     * range. If possible, we implement this using bounded cursors to prevent false conflicts with
```

Comment:
I don't think "false conflicts" is accurate here, I'd like to rephrase this a bit but you can edit my rephrase as you see fit.

"Truncate does not require keys actually exist so that applications can discard parts of the object's name space without knowing exactly what records currently appear in the object. When possible use bounded cursors to position the start and stop cursors, if they aren't available then use search-near. search-near is sub optimal as it may return prepare conflicts outside of the truncate key range as it will walk beyond the end key."

Reply:
Sure, I will use this one too. The definition of "false conflict" is an interesting one, but if there is any confusion or a potential for misunderstanding, no need to mention it in the comment. (I used the term, as from the user's perspective, there should be no conflict.)